Pusha imagenes base y jupyter a Dockerhub desde GitHub Actions

devarops opened the pull request:
[Github pusha imagenes base y jupyter a Dockerhub](https://app.gitkraken.com/glo/board/5e711b4e31df0b001242edf6/card/5e727ae41332eb00111285b4)

devarops committed:
Consignación inicial

devarops committed:
Corrige configuración de GitHub Actions

devarops committed:
Especifica Makefile usado para construir los Dockerfiles

devarops committed:
Define onjetivos _phonies_

devarops committed:
Sube imágenes a Docker Hub

devarops committed:
Sube imágenes a Docker Hub sólo en develop, master

En push y pull request sólo contruye imágenes y corre pruebas sin pushar a Docker Hub

devarops committed:
Corre cada Action en la rama correspondiente

devarops committed:
Mejora nómbres hacíendolos más cortos
porque así es más fácil de leer en GitHub

islasfernando commented on .github/workflows/develop.yml:
```diff
@@ -0,0 +1,31 @@
+      env:
+        DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+        DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+      run: echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
```
no estoy seguro si deberían de ser `"${DOCKER_PASSWORD}"` y `"${DOCKER_USERNAME}"` de acuerdo con la guía de estilo Shell?

devarops replied:
Listo

islasfernando commented on .github/workflows/master.yml:
```diff
@@ -0,0 +1,31 @@
+      env:
+        DOCKER_USERNAME: ${{ secrets.DOCKER_USERNAME }}
+        DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
+      run: echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
```
Acá lo mismo

devarops replied:
Listo

islasfernando commented on _Dockerfile.footer_jupyter:
```diff
@@ -0,0 +1,12 @@
+    'utm' \
+        && \
+    conda clean --all -f -y && \
+    fix-permissions $CONDA_DIR
```
sirve `${CONDA_DIR}` ?

devarops replied:
Listo

islasfernando commented on _Dockerfile.footer_jupyter:
```diff
@@ -0,0 +1,12 @@
+    'pandas' \
+    'utm' \
+        && \
+    conda clean --all -f -y && \
```
el `-f` y el `-y` tienen forma larga? si es así, pofavor agregar

devarops replied:
Listo

islasfernando commented on .github/workflows/develop.yml:
```diff
@@ -0,0 +1,31 @@
+    - name: Construye Dockerfiles
+      run: make --makefile=Makefile.build dockerfiles
+    - name: Construye imagen Base
+      run: docker build --tag islasgeci/base:latest --file Dockerfile.base .
+    - name: Construye imagen Jupyter
+      run: docker build --tag islasgeci/jupyter:latest --file Dockerfile.jupyter .
```
Ya no habrá tags particulares entonces?

devarops replied:
Sí, eso los hace el contenedor [`tag_docker_images`](https://github.com/IslasGECI/tag_docker_images)

islasfernando commented on Makefile:
```diff
@@ -0,0 +1,25 @@
+
+test_tidyverse:
+	R -e "library('tidyverse')"
+
```
valdrá la pena alguna prueba para `python`? o pruebas independientes, que solo corran en base o solo en jupyter y que cada imágen tenga sus pruebas?

devarops replied:
Para Python podríamos probar que los módulos numpy, pandas y matplotlib están instalados. ¿Se te ocurre algún otro módulo básico de Python?

islasfernando replied:
Creo que está bien con esos, supongo que han de traer dependencias que también se van a instalar.

devarops replied:
OK. Entonces agregaré una prueba para eso

islasfernando commented on _Dockerfile.header_base:
```diff
@@ -0,0 +1 @@
+FROM rocker/r-ubuntu:18.04
```
Entonces la base está más orientado a R? El otro día quise usar base en un repo pero no tienía las paqueterías básicas de python (numpy, pandas, matplotlib etc.)

devarops replied:
No, la base no está más orientada a R. Es un accidente que Anaconda ya traiga numpy, pandas y matplotlib instalados.
¿Cuáles son los módulos básicos de Python además de numpy, pandas, matplotlib que ya vienen en Jupyter?

islasfernando commented on _Dockerfile.footer_base:
```diff
@@ -0,0 +1,12 @@
+    rm /usr/bin/python && \
+    ln --symbolic /usr/bin/python3 /usr/bin/python
+RUN pip install \
+    goodtables \
```
Tal vez probar goodtables sea la opción de la prueba de `python` con el `test.csv`?

devarops replied:
Prefiero hacer una prueba específica para Python

devarops committed:
Nombra las variables con minúsculas para seguir guía de estilo

devarops committed:
Usa llaves en variable de Shell

devarops committed:
Usa versión larga de las opciones de `conda clean`

devarops committed:
Verifica que Python sea versión 3

devarops committed:
Mueve la definición del vínculo de pytnon->python3 del header al footer

Moví  `ln --symbolic /usr/bin/python3 /usr/bin/python` del _header_ al _footer_ para dejar ambos _headers_ con una única línea

devarops committed:
Verifica que los módulos matplotlib, numpy y pandas estén instalados

devarops committed:
Elimina versión default de Python para sustituirla por python3

devarops committed:
Instala módulos de Python faltantes: matplotlib, numpy, pandas

devarops commented on Makefile:
```diff
@@ -0,0 +1,33 @@
+test_python:
+	[ "$$(python --version | cut --characters=1-8)" == "Python 3" ]
+
+test_python_modules:
+	pip freeze | grep matplotlib
+	pip freeze | grep numpy
+	pip freeze | grep pandas
```
@islasfernando Aquí verifico que estos tres módulos estén instalados

devarops commented on Makefile:
```diff
@@ -0,0 +1,33 @@
+.PHONY: clean tests test_python test_python_modules test_pythontex test_queries test_tidyverse
+
+test_python:
+	[ "$$(python --version | cut --characters=1-8)" == "Python 3" ]
```
@islasfernando Aquí verifico que la versión de Python sea 3
